File: src/main/java/xyz/nkomarn/composter/entity/Player.java
package xyz.nkomarn.composter.entity;

import kyta.composter.world.ChunkPos;
import net.kyori.adventure.text.Component;
import org.jetbrains.annotations.NotNull;
import xyz.nkomarn.composter.command.CommandSource;
import xyz.nkomarn.composter.entity.tracker.EntityTracker;
import xyz.nkomarn.composter.network.Connection;
import xyz.nkomarn.composter.network.protocol.packet.play.ClientboundChunkDataPacket;
import xyz.nkomarn.composter.network.protocol.packet.play.ClientboundChunkOperationPacket;
import xyz.nkomarn.composter.network.protocol.packet.s2c.ClientboundChatPacket;
import xyz.nkomarn.composter.world.World;

import java.util.HashSet;
import java.util.Set;

public final class Player extends Entity implements CommandSource {
    private static final int VIEW_DISTANCE = 16;
    public static final double DEFAULT_STANCE = 67.240000009536743;

    private final Connection connection;
    private final String username;
    private final EntityTracker tracker;
    private final Set<ChunkPos> visibleChunks;
    private boolean crouching;
    private boolean onGround;
    private long lastDigStartTime;

    //TODO crouching support

    public Player(World world, Connection connection, String username) {
        super(world);
        this.connection = connection;
        this.username = username;
        this.tracker = new EntityTracker(this);
        this.visibleChunks = new HashSet<>();
    }

    public Connection connection() {
        return connection;
    }

    public String getUsername() {
        return username;
    }

    @Override
    public String getName() {
        return getUsername();
    }

    @Override
    public void sendMessage(@NotNull Component message) {
        connection.sendPacket(new ClientboundChatPacket(message));
    }

    /*
    public void teleport(@NotNull Location location) {
        this.location = location;
        updateVisibleChunks();
        // updateLocation();
    }
     */

    public boolean isCrouching() {
        return crouching;
    }

    public void setCrouching(boolean crouching) {
        this.crouching = crouching;
    }

    public boolean isOnGround() {
        return onGround;
    }

    public void setOnGround(boolean onGround) {
        this.onGround = onGround;
    }

    public double getStance() {
        return DEFAULT_STANCE;
    }

    public long getLastDigStartTime() {
        return lastDigStartTime;
    }

    public void setLastDigStartTime(long lastDigStartTime) {
        this.lastDigStartTime = lastDigStartTime;
    }

    @Override
    public void tick(long currentTick) {
        super.tick(currentTick);
        tracker.tick(currentTick);

        /*
         * if this player has moved, update their visible chunks.
         */
        // if (tracker.hasChangedBlock()) {
        updateVisibleChunks();
        // }

        // this.session.sendPacket(new KeepAliveBiPacket());
    }

    public void updateVisibleChunks() {
        var currentBlock = getBlockPos();
        var currentChunk = new ChunkPos(currentBlock);
        int currentChunkX = currentChunk.getX();
        int currentChunkZ = currentChunk.getZ();

        /*
         * send chunk packets for chunks that the player should
         * be able to see from their current position.
         */
        var pendingUnload = new HashSet<>(visibleChunks);

        for (var x = (currentChunkX - VIEW_DISTANCE); x < (currentChunkX + VIEW_DISTANCE); x++) {
            for (var z = (currentChunkZ - VIEW_DISTANCE); z < (currentChunkZ + VIEW_DISTANCE); z++) {
                var pos = new ChunkPos(x, z);

                /*
                 * if the chunk is currently visible, don't
                 * send another chunk data packet to the client.
                 */
                if (visibleChunks.contains(pos)) {
                    pendingUnload.remove(pos);
                    continue;
                }

                /*
                 * if the chunk hasn't been loaded on the server
                 * side, don't send it to the client yet.
                 */
                var chunk = getWorld().getLoadedChunk(pos);
                if (chunk == null) continue;

                connection.sendPacket(new ClientboundChunkOperationPacket(x, z, true));
                connection.sendPacket(new ClientboundChunkDataPacket(chunk));
                visibleChunks.add(pos);
            }
        }

        /*
         * unload chunks that are no longer visible.
         */
        pendingUnload.forEach(pos -> {
            visibleChunks.remove(pos);
            connection.sendPacket(new ClientboundChunkOperationPacket(pos.getX(), pos.getZ(), false));
        });
    }
}

File: src/main/java/xyz/nkomarn/composter/entity/Player.kt
package xyz.nkomarn.composter.entity

import kyta.composter.container.BasicContainer
import kyta.composter.container.menu.Menu
import kyta.composter.container.menu.PlayerInventoryMenu
import kyta.composter.entity.EntityType
import kyta.composter.item.ItemStack
import kyta.composter.network.Connection
import kyta.composter.protocol.Packet
import kyta.composter.protocol.packet.play.ClientboundAddPlayerPacket
import kyta.composter.protocol.packet.play.ClientboundChatMessagePacket
import kyta.composter.protocol.packet.play.ClientboundChunkDataPacket
import kyta.composter.protocol.packet.play.ClientboundChunkOperationPacket
import kyta.composter.protocol.packet.play.ClientboundSetContainerContentPacket
import kyta.composter.protocol.packet.play.GenericPlayerActionPacket
import kyta.composter.world.ChunkPos
import kyta.composter.world.World
import net.kyori.adventure.text.Component
import xyz.nkomarn.composter.entity.tracker.EntityTracker
import java.util.function.Consumer

class Player(
    world: World,
    val connection: Connection,
    val username: String,
) : Entity(world, EntityType.PLAYER) {
    private val visibleChunks = mutableSetOf<ChunkPos>()
    override val dimensions = 0.6 to 1.8
    val inventory = BasicContainer(36)
    val armorContainer = BasicContainer(4)
    var selectedHotbarSlot = 0
    var cursorItem: ItemStack = ItemStack.EMPTY

    var inventoryMenu = PlayerInventoryMenu(this, inventory, armorContainer)
    var currentMenu: Menu? = null

    val entityTracker = EntityTracker(this)
    val stance = 67.240000009536743
    var isCrouching = false
    var isOnGround = false
    var lastDigStartTime = 0L

    var menuCounter = 0

    fun swingArm() {
        entityTracker.broadcast(GenericPlayerActionPacket(id, GenericPlayerActionPacket.Action.SWING_ARM))
    }

    override fun createAddEntityPacket(): Packet {
        return ClientboundAddPlayerPacket(this)
    }

    fun sendMessage(message: Component) {
        connection.sendPacket(ClientboundChatMessagePacket(message))
    }

    override fun tick(currentTick: Long) {
        super.tick(currentTick)
        entityTracker.tick(currentTick)
        updateVisibleChunks()

        /* send menu updates */
        currentMenu?.tick(currentTick)
            ?: inventoryMenu.tick(currentTick)
    }

    private fun updateVisibleChunks() {
        val currentBlock = getBlockPos()
        val (currentChunkX, currentChunkZ) = ChunkPos(currentBlock)

        /*
         * send chunk packets for chunks that the player should
         * be able to see from their current position.
         */
        val pendingUnload = HashSet(visibleChunks)
        for (x in currentChunkX - VIEW_DISTANCE until currentChunkX + VIEW_DISTANCE) {
            for (z in currentChunkZ - VIEW_DISTANCE until currentChunkZ + VIEW_DISTANCE) {
                val pos = ChunkPos(x, z)

                /*
                 * if the chunk is currently visible, don't
                 * send another chunk data packet to the client.
                 */
                if (visibleChunks.contains(pos)) {
                    pendingUnload.remove(pos)
                    continue
                }

                /*
                 * if the chunk hasn't been loaded on the server
                 * side, don't send it to the client yet.
                 */
                val chunk = world.chunks.getLoadedChunk(pos) ?: continue
                connection.sendPacket(ClientboundChunkOperationPacket(pos, ClientboundChunkOperationPacket.Mode.LOAD))
                connection.sendPacket(ClientboundChunkDataPacket(pos, chunk))
                visibleChunks.add(pos)
            }
        }

        /*
         * unload chunks that are no longer visible.
         */
        pendingUnload.forEach(Consumer { pos: ChunkPos ->
            visibleChunks.remove(pos)
            connection.sendPacket(ClientboundChunkOperationPacket(pos, ClientboundChunkOperationPacket.Mode.UNLOAD))
        })
    }

    private companion object {
        const val VIEW_DISTANCE = 16
    }
}
